### ISSUE TITLE
Copy chart image file directly to the clipboard

### ISSUE THREAD

**Issue by cary-sas** [highcharts/highcharts#22167]
# Description of the feature

Hope to add an option "Copy chart to clipboard " on highcharts-contextmenu.  And it also support the chart running on a HTTP server.

![image](https://github.com/user-attachments/assets/d2364d55-f34a-49c5-b9f6-dd2098d52b4b)

# Library related to the feature

offline-exporting.js
exporting.js

# Proof of Concept/Live example for the feature

----

**You can vote for this feature by adding a thumbs-up reaction to this post.**

**Comment by hubertkozik** [comment]
Hi @cary-sas! Thank you for sharing your idea.

You can create a custom button in exporting context menu by setting [exporting.buttons.contextButton.menuItems](https://api.highcharts.com/highcharts/exporting.buttons.contextButton.menuItems). Then you can fire the desired function on click, e.g. you can get the chart SVG copy and copy it to the clipboard. You can check the demo below to see an example.

**Live demo:** https://jsfiddle.net/BlackLabel/mdr5eg72/show
https://jsfiddle.net/BlackLabel/mdr5eg72/

Currently, there is no possibility of getting the chart image copied to the clipboard. You can use a third-party plugin to achieve this.

**Comment by hubertkozik** [comment]
Yes, currently, the chart image cannot be copied to the clipboard. However, you can use some third-party plugin to achieve this.

**Comment by hubertkozik** [comment]
@cary-sas I did some more investigation about your idea and there's an experimental feature in the **ClipboardItem API** that supports SVG images. As of November 2024, this feature is implemented in Chrome, Edge, and Opera. You can check the compatibility details here: [Can I Use - ClipboardItem API (image/svg+xml)](https://caniuse.com/mdn-api_clipboarditem_type_image_svg_xml).

I followed [Google's blog post](https://developer.chrome.com/blog/svg-support-for-async-clipboard-api) and implemented a simple demo for you. You can check it below.

**Live demo:** https://jsfiddle.net/BlackLabel/9qx3oufd/

**Comment by lebesnec** [comment]
this is what we use to convert the graph to png and put it in the clipboard :

```
function exportToClipboard(graph) {
    const image = new Image();
    image.onload = () => {
      const canvas = document.createElement('canvas');
      canvas.width = 1024;
      canvas.height = 768;
      const context = canvas.getContext('2d');
      context.drawImage(image, 0, 0);
      canvas.toBlob((blob) => navigator.clipboard.write([ new ClipboardItem({ 'image/png': blob }) ]), 'image/png');
    };

    const svg = new Blob([ chart.getSVG() ], { type: 'image/svg+xml;charset=utf-8' });
    image.src = URL.createObjectURL(svg);
}
```

You could put the SVG in the clipboard directly but a lot of application don't support this format when pasted.

**Comment by xPaw** [comment]
This indeed would be great to have. Highcharts v12.3.0 refactored exporting module to make local exporting the default, so it is definitely possible to add now.

**Comment by xPaw** [comment]
The following works fine. It will be easier to integrate directly in the exporting code, because it can reuse functions like `imageToDataURL` and `localExport` directly, it just needs to call `toBlob` on the canvas and write it to the clipboard (with error handling of course).

```js
copyChartToClipboard:
{
	text: 'Copy PNG to clipboard',
	onclick()
	{
		const that = this;
		const type = 'image/png';

		function getBlob()
		{
			return new Promise( ( resolve ) =>
			{
				// @ts-ignore
				const chartSVG = window.Highcharts.Exporting.svgToDataURL( that.exporting.getSVGForExport( that.exporting.options ) );

				const image = new Image();
				image.crossOrigin = 'anonymous';
				image.src = chartSVG;

				image.addEventListener( 'load', () =>
				{
					const canvas = document.createElement( 'canvas' );
					const ctx = canvas.getContext( '2d' );

					canvas.height = image.height; // * scale
					canvas.width = image.width; // * scale
					ctx.drawImage( image, 0, 0, canvas.width, canvas.height );

					canvas.toBlob( ( blob ) =>
					{
						resolve( blob );
					}, type );
				} );
			} );
		}

		navigator.clipboard.write( [
			new ClipboardItem( { [ type ]: getBlob() } )
		] );
	},
},
```

EDIT: Changed to write promise into clipboard item
